### 作業
目前已經學過許多的模型，相信大家對整體流程應該比較掌握了，這次作業請改用**手寫辨識資料集**，步驟流程都是一樣的，請試著自己撰寫程式碼來完成所有步驟

- [GradientBoostingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html)

In [3]:
from sklearn import datasets, metrics
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split

# read
digits = datasets.load_digits()

# split
x_train, x_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.2, random_state=42)

# model
gbc = GradientBoostingClassifier().fit(x_train, y_train)

# predict
y_pred = gbc.predict(x_test)

# score
print(f'accurancy: {metrics.accuracy_score(y_test, y_pred)}')
print(f'accurancy: {metrics.accuracy_score(y_pred, y_test)}')

accurancy: 0.9694444444444444
accurancy: 0.9694444444444444


In [11]:
# 改一些參數試試看
# read
digits = datasets.load_digits()

# split
x_train, x_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.2, random_state=42)

# model
# gbc = GradientBoostingClassifier(loss=’deviance’, learning_rate=0.1, n_estimators=100, subsample=1.0, criterion=’friedman_mse’, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_depth=3, min_impurity_decrease=0.0, min_impurity_split=None, init=None, random_state=None, max_features=None, verbose=0, max_leaf_nodes=None, warm_start=False, presort=’auto’, validation_fraction=0.1, n_iter_no_change=None, tol=0.0001).fit(x_train, y_train)
# 上面是完整的
# gbc = GradientBoostingClassifier(learning_rate=0.2).fit(x_train, y_train)
# gbc = GradientBoostingClassifier(learning_rate=0.3).fit(x_train, y_train)
# gbc = GradientBoostingClassifier(learning_rate=0.3, n_estimators=200).fit(x_train, y_train)
gbc = GradientBoostingClassifier(learning_rate=0.1, n_estimators=200).fit(x_train, y_train)



# predict
y_pred = gbc.predict(x_test)

# score
print(f'accurancy: {metrics.accuracy_score(y_test, y_pred)}')
print(f'accurancy: {metrics.accuracy_score(y_pred, y_test)}')

accurancy: 0.9722222222222222
accurancy: 0.9722222222222222


- 改參數過程
- [參考調整參數統整](https://blog.csdn.net/brucewong0516/article/details/78700402)
    - learning rate=0.2 ---> 沒變。
    - learning rate=0.3 ---> 提升大約0.03至0.9722
    - learning rate=0.3, n_estimators=200 ---> 同上, 沒變
    - 一直推到400都沒變。換別的
    - learning rate=0.1, n_estimators=200 ---> 0.9722, 學習率不要太高
    